libs/vkd3d/command.c

```diff
@@ -2692,6 +2696,339 @@ static void d3d12_command_list_handle_active_queries(struct d3d12_command_list *
     }
 }
 
+int vkd3d_compare_pending_query(const void* query_a, const void* query_b)
+{
+    const struct vkd3d_active_query *a = query_a;
+    const struct vkd3d_active_query *b = query_b;
+
+    // Sort by D3D12 heap since we need to do one compute dispatch per buffer
+    if (a->heap < b->heap) return -1;
+    if (a->heap > b->heap) return 1;
+
+    // Sort by Vulkan query pool and index to batch query resolves
+    if (a->vk_pool > b->vk_pool) return -1;
+    if (a->vk_pool < b->vk_pool) return 1;
+
+    return (int)(a->vk_index - b->vk_index);
+}
+
+static size_t get_query_heap_stride(D3D12_QUERY_HEAP_TYPE heap_type)
+{
+    if (heap_type == D3D12_QUERY_HEAP_TYPE_PIPELINE_STATISTICS)
+        return sizeof(D3D12_QUERY_DATA_PIPELINE_STATISTICS);
+
+    if (heap_type == D3D12_QUERY_HEAP_TYPE_SO_STATISTICS)
+        return sizeof(D3D12_QUERY_DATA_SO_STATISTICS);
+
+    return sizeof(uint64_t);
+}
+
+static bool d3d12_command_list_gather_pending_queries(struct d3d12_command_list *list)
+{
+    VkDeviceSize resolve_buffer_size, resolve_buffer_stride, ssbo_alignment, entry_buffer_size;
+    const struct vkd3d_vk_device_procs *vk_procs = &list->device->vk_procs;
+    struct vkd3d_scratch_allocation resolve_buffer, entry_buffer;
+    VkDescriptorBufferInfo dst_buffer, src_buffer, map_buffer;
+    struct vkd3d_query_gather_info gather_pipeline;
+    const struct vkd3d_active_query *src_queries;
+    unsigned int i, j, k, workgroup_count;
+    uint32_t resolve_index, entry_offset;
+    struct vkd3d_query_gather_args args;
+    VkWriteDescriptorSet vk_writes[3];
+    VkMemoryBarrier vk_barrier;
+    VkDescriptorSet vk_set;
+    bool result = false;
+
+    struct dispatch_entry
+    {
+        struct d3d12_query_heap *heap;
+        uint32_t virtual_query_count;
+        uint32_t unique_query_count;
+        VkDeviceSize resolve_buffer_offset;
+        VkDeviceSize resolve_buffer_size;
+    };
+    
+    struct dispatch_entry *dispatches = NULL;
```

Comment:
Any way to avoid having to allocate fresh arrays every time?

Reply:
Could technically store the arrays in command allocator, but I can see why it gets very hairy.

Reply:
Can do, it's just annoying, and we shouldn't hit this more than once per frame anyway.

Reply:
Could just mark this as TODO at least so that we have a potential optimization avenue if games are dumb.

Reply:
Done.